documentation/projects/proposals/search_relevancy_sandbox/20230518-implementation_plan_staging_index_rapid_iteration.md

```diff
@@ -0,0 +1,380 @@
+will need to be created for each Elasticsearch cluster. The connections will be
+named `elasticsearch_production` and `elasticsearch_staging` and reference each
+respective environment. These will be created by hand in the Airflow UI. For
+local development, defaults will be added to the
+[`env.template` file](https://github.com/WordPress/openverse/blob/3fcce5ade2165955db5bbcb4f679257b3260547b/catalog/env.template).
+
+### Infrastructure
+
+<!-- Describe any infrastructure that will need to be provisioned or modified. In particular, identify associated potential cost changes. -->
+
+No infrastructure changes should be necessary for this implementation, beyond
+the addition of the `elasticsearch` provider package and the Airflow connections
+described above.
+
+### Other projects or work
+
+<!-- Note any projects this plan is dependent on. -->
+
+This work does not depend on any other projects and can be implemented at any
+time. In order to be utilized however, a query parameter on the API will need to
+be used for determining which index to query. This has been implemented in
+specifying which index to use for a given query. This has been implemented in
+#2073.
+
+## Outlined Steps
+
+<!-- Describe the implementation step necessary for completion. -->
+
+Once the prerequisites above have been filled, the DAGs can be created. The DAGs
+will be named `create_new_es_index_{environment}`, and will have the parameters
+and steps described below. A dynamic DAG generation function will be used to
+create one DAG per environment (`staging` and `production`). It will have a
+schedule of `None` so that it is only run when triggered. It should also have
+`max_active_runs` set to `1` so that only one instance of the DAG can be running
+at a time.
+
+### Parameters
+
+1. `media_type`: The media type for which the index is being created. Presently
+   this would only be `image` or `audio`.
+2. `index_config`: A JSON object containing the configuration for the new index.
+   The values in this object will be merged with the existing configuration,
+   where the value specified at a leaf key in the object will override the
+   existing value (see [Merging policy](#merging-policy) below). This can also
+   be the entire index configuration, in which case the existing configuration
+   will be replaced entirely (see `override_config` parameter below).
+3. `index_suffix`: (Optional) The name suffix of the new index to create. This
+   will be a string, and will be used to name the index in Elasticsearch of the
+   form `{media_type}-{index_suffix}`. If not provided, the suffix will be a
+   timestamp of the form `YYYYMMDDHHMMSS`.
+4. `source_index`: (Optional) The existing index on Elasticsearch to use as the
+   basis for the new index. If not provided, the index aliased to `media_type`
+   will be used (e.g. `image` for the `image` media type).
```

Comment:
Requesting changes for clarifications around the `source_index`, if the default index is assumed to come from the data refresh process I believe is worth pointing it out.

Reply:
I answered this a bit above but I'll be sure to add some extra context for where this might be coming from in staging!